### PR TITLE
[FLINK-15961][table-planner][table-planner-blink] Introduce Python Physical Correlate RelNodes which are containers for Python TableFunction

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request introduce Python Physical Correlate RelNodes which are containers for Python TableFunction*

## Brief change log

  - *Add Physical Python Correlate Rule `BatchExecPythonCorrelateRule` ,`StreamExecPythonCorrelateRule` and `DataStreamPythonCorrelateRule`*
  - *Add Physical Python Correlate Node `DataStreamPythonCorrelate`, `StreamExecPythonCorrelate` and `BatchExecPythonCorrelate`*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added ut case testCorrelatePythonTableFunction*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecCorrelate.scala [hunk 1]
```diff
@@ -21,59 +21,37 @@ import org.apache.flink.api.dag.Transformation
...
-
-import java.util
-
-import scala.collection.JavaConversions._
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rex.{RexNode, RexProgram}
 
 /**
-  * Flink RelNode which matches along with join a user defined table function.
+  * Flink RelNode which matches along with join a java user defined table function.
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecCorrelate.scala]: java => Java/Scala

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/BatchExecPythonCorrelateRule.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+		if (right instanceof FlinkLogicalTableFunctionScan) {
+			// right node is a python table function
+			FlinkLogicalTableFunctionScan scan = (FlinkLogicalTableFunctionScan) right;
+			return PythonUtil.isPythonCall(scan.getCall());
+		} else if (right instanceof FlinkLogicalCalc) {
+			// a filter is pushed above the table function
+			FlinkLogicalCalc calc = (FlinkLogicalCalc) right;
+			RelSubset input = (RelSubset) calc.getInput();
+			FlinkLogicalTableFunctionScan scan = (FlinkLogicalTableFunctionScan) input.getOriginal();
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/BatchExecPythonCorrelateRule.java]: We can't cast the original node to `FlinkLogicalTableFunctionScan`.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/BatchExecPythonCorrelateRule.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	 */
+	private static class BatchExecPythonCorrelateFactory {
+		private final RelNode correlateRel;
+		private final FlinkLogicalCorrelate correlate;
+		private final RelTraitSet traitSet;
+		private final RelNode convInput;
+		private final RelNode right;
+
+		BatchExecPythonCorrelateFactory(RelNode rel) {
+			this.correlateRel = rel;
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/BatchExecPythonCorrelateRule.java]: Remove this member?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecPythonCorrelateRule.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	public RelNode convert(RelNode relNode) {
+		StreamExecPythonCorrelateFactory factory = new StreamExecPythonCorrelateFactory(relNode);
+		return factory.convertToCorrelate();
+	}
+
+	/**
+	 * The factory is responsible to creating {@link StreamExecPythonCorrelate}.
+	 */
+	private static class StreamExecPythonCorrelateFactory {
+		private final RelNode correlateRel;
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecPythonCorrelateRule.java]: Remove this member.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecPythonCorrelateRule.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	@Override
+	public RelNode convert(RelNode relNode) {
+		StreamExecPythonCorrelateFactory factory = new StreamExecPythonCorrelateFactory(relNode);
+		return factory.convertToCorrelate();
+	}
+
+	/**
+	 * The factory is responsible to creating {@link StreamExecPythonCorrelate}.
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/StreamExecPythonCorrelateRule.java]: responsible to => responsible for

**Comment by HuangXingBo** [comment]
Thanks a lot for @hequn8128 review, I have addressed the comments at the latest commit.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecCorrelateBase.scala [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    */
+  def copy(
+      traitSet: RelTraitSet,
+      newChild: RelNode,
+      projectProgram: Option[RexProgram],
+      outputType: RelDataType): RelNode
+
+  override def explainTerms(pw: RelWriter): RelWriter = {
+    val rexCall = scan.getCall.asInstanceOf[RexCall]
+    val sqlFunction = rexCall.getOperator.asInstanceOf[TableSqlFunction]
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecCorrelateBase.scala]: The test failed due to this line. We can remove this line as it has never been used.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecCorrelateBase.scala [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    */
+  def copy(
+      traitSet: RelTraitSet,
+      child: RelNode,
+      projectProgram: Option[RexProgram],
+      outputType: RelDataType): RelNode
+
+  override def explainTerms(pw: RelWriter): RelWriter = {
+    val rexCall = scan.getCall.asInstanceOf[RexCall]
+    val sqlFunction = rexCall.getOperator.asInstanceOf[TableSqlFunction]
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecCorrelateBase.scala]: ditto

**Comment by HuangXingBo** [comment]
> @HuangXingBo Thanks a lot for the update. The test failed due to the problems in `BatchExecCorrelateBase` and `StreamExecCorrelateBase`. See the detailed comments below.
> 
> As for the last commit that adding implementations for the Python Correlate RelNode, maybe it's better to add the commit later. Because in this PR, there are no ways to write tests to cover the implementation.
> 
> What do you think?

Thanks a lot for @hequn8128 , I will move the last commit to FLINK-15972.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamCorrelateBase.scala [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.calcite.rex.{RexCall, RexNode}
+import org.apache.flink.table.functions.utils.TableSqlFunction
+import org.apache.flink.table.plan.nodes.CommonCorrelate
+import org.apache.flink.table.plan.nodes.logical.FlinkLogicalTableFunctionScan
+import org.apache.flink.table.plan.schema.RowSchema
+
+/**
+  * Base RelNode for data stream correlate.
+  */
+abstract class DataStreamCorrelateBase(
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamCorrelateBase.scala]: Hi @hequn8128, thanks for the notice. Sure, in such cases we can keep classes as Scala classes as a short term solution. But we should eliminate those shortcomings in the near future. E.g. by adding a refactoring commit that unblocks new code from implementing it in Java. In any case we should create a follow-up issue and find someone to address this.

**Comment by hequn8128** [comment]
@flinkbot run travis

**Comment by hequn8128** [comment]
@HuangXingBo Hi, the jira id in the commit log is wrong. Would be great if you can check on this next time.
Merging...

**Pull request closed without merging**